fix(op-batcher): accurate pending DA metrics

joshklop opened the pull request:
Closes #14559. Depends on #16773

joshklop committed:
op-batcher: always use blobs while throttling

geoknee commented on op-batcher/batcher/channel_manager.go:
```diff
@@ -183,13 +184,16 @@ func (s *channelManager) handleChannelInvalidated(c *channel) {
 			}
 		}
 	}
+	c.metr.RecordChannelTimedOut(c.ID(), invalidatedChannelLengths)
```
I'm not sure we want to move this here. If this function executes, the channel was invalidated -- there are multiple possible reasons why, and only one of them is that the channel timed out.

geoknee commented on op-batcher/metrics/metrics.go:
```diff
@@ -69,6 +69,25 @@ type Metricer interface {
 	PendingDABytes() float64
 }
 
+type BlockWithSizeEstimate struct {
+	*types.Block
+	daSize  uint64
+	rawSize uint64
+}
+
+func NewBlockWithSizeEstimate(inner *types.Block) *BlockWithSizeEstimate {
+	return &BlockWithSizeEstimate{
+		Block: inner,
+	}
+}
+
+func (b *BlockWithSizeEstimate) Estimate() (uint64, uint64) {
+	if b.daSize == 0 {
+		b.daSize, b.rawSize = estimateBatchSize(b.Block)
+	}
+	return b.daSize, b.rawSize
```
I've always been a bit uneasy about the metrics package being so central to the business logic of the batcher. Putting this new type here pushes further in that direction -- did you consider putting it in `types`?

joshklop replied:
Agreed, will change course.

geoknee commented on op-batcher/metrics/metrics.go:
```diff
@@ -386,17 +405,17 @@ func (m *Metrics) RecordChannelClosed(id derive.ChannelID, numPendingBlocks int,
+	daSize, rawSize := block.Estimate()
 	m.pendingBlocksBytesCurrent.Add(-1.0 * float64(rawSize))
-	atomic.AddInt64(&m.pendingDABytes, -1*int64(daSize))
+	atomic.AddInt64(&m.pendingDABytes, -1*int64(daSize)+int64(size))
```
The problem we are trying to solve here is to have a stable idea of the pending data to be written to L1, so that we can use that to drive throttling accurately. The main problem at the moment is that `pendingDABytes` can fluctuate a lot depending on where blocks are in the batcher's data pipeline. `pendingDABytes` goes to zero when all blocks move from the pending queue into a channel. We effectively need to change the definition of "pending DA bytes" from covering only blocks not yet in channels to all data which has yet to make it on L1. 

A very basic solution to this would just be to keep `pendingDABytes` in sync with the estimated DA size of all blocks in the channel manager. (ignoring the block cursor and channels altogether). This would be ignoring the reduction in DA volume we get by compressing transactions _together_ in channels, but that assumption is made elsewhere in the protocol and might be an acceptable tradeoff against the benefits of KISS. `pendingDABytes` would then only decrease when those blocks become safe and are pruned. There is an added benefit here that we wouldn't need to hook into channels invalidating (timing out or being replaced with channels of a different DA type) at all because it would be irrelevant. The only events to hook into then are enqueueing and dequeueing from the channel managers block queue. 

I'm finding the usage of `daSize`, `rawSize` and just `size` a little difficult to reason about, along with having to increase and decrease `pendingDABytes` when we put a block into a channel. So I'd like to consider simpler alternatives if possible.

joshklop replied:
Trading off some precision for simplicity seems worthwhile. I'll consider other options.

geoknee reviewed (commented):
Looking good so far but left some comments.

Because we are changing the definition of `pendingDABytes`, we will need to call this out in the release notes and make changes to metrics dashboards as appropriate.

joshklop committed:
op-batcher: improve pending DA metrics accuracy

Rather than subtracting a block's estimated DA size it is added to a channel,
we subtract when it is marked as safe.

geoknee commented on op-batcher/metrics/metrics.go:
```diff
@@ -40,7 +40,7 @@ type Metricer interface {
 	RecordChannelOpened(id derive.ChannelID, numPendingBlocks int)
 	RecordL2BlocksAdded(l2ref eth.L2BlockRef, numBlocksAdded, numPendingBlocks, inputBytes, outputComprBytes int)
 	RecordL2BlockInPendingQueue(block *types.Block)
-	RecordL2BlockInChannel(block *types.Block)
+	RecordL2BlockDequeued(block *types.Block)
```
`RecordL2BlockDequeued` is only called in one place, when we dequeue a block (pruning because it safe).

`RecordL2BlockInPendingQueue` is called in _two_ places. Once when we add it (loading from sequencer), which makes sense, and once when we rewind the block cursor:

https://github.com/ethereum-optimism/optimism/blob/88ef72cd783f11aba5035f4bcc0e21fa503a6e5e/op-batcher/batcher/channel_manager.go#L140-L147

We were doing this before because pending meant that it was not yet in a channel -- rewinding happens when we take blocks back out of a channel and use the block cursor to mark them as ready to be processed into a channel again.

Now that pending means "not yet safe" instead of "not yet in a channel" I think we want to remove this block of code.

joshklop replied:
Nice catch, thanks.

geoknee commented on op-batcher/batcher/channel_manager_test.go:
```diff
@@ -566,11 +565,13 @@ func TestChannelManager_PruneBlocks(t *testing.T) {
 			m.blockCursor = tc.initialBlockCursor
 			if tc.expectedQ != nil {
 				m.PruneSafeBlocks(tc.numChannelsToPrune)
+				require.Negative(t, metr.PendingDABytes()) // Pruning will subtract from zero.
```
I wonder if it would be less surprising here to mirror the metric _increase_ we would usually have when adding blocks through the normal path, before invoking the pruning method. In this way we would avoid metrics going negative. Seems a bit strange to assert on a condition which we never want to arise in production.

joshklop replied:
Fair point. As written, this is necessary since the test injects the initial queue outside the public api (which would mandate `AddL2Block`). I'll change the surrounding lines to use `AddL2Block` so the metrics are properly updated and give us a more natural test.

geoknee commented on op-batcher/metrics/metrics.go:
```diff
@@ -393,7 +393,7 @@ func (m *Metrics) RecordL2BlockInPendingQueue(block *types.Block) {
 }
 
-func (m *Metrics) RecordL2BlockInChannel(block *types.Block) {
+func (m *Metrics) RecordL2BlockDequeued(block *types.Block) {
```
We should update the comment for the underlying variable updated to accurately reflect the new meaning of "pending":

https://github.com/ethereum-optimism/optimism/blob/c12539a8af3a05fd6b4a29846d82f759613735f7/op-batcher/metrics/metrics.go#L311-L315

We can reference the approximate nature of the variable as well, by explaining how we estimate the DA footprint.

geoknee replied:
It occurs to me now that we need to be careful about consistency around "pending". I think there are other places where we use this to mean "not yet in a channel". I just want to avoid confusion, let's discuss how to resolve those.

joshklop replied:
Good point. I have manually checked all occurrences of the word "pending" in the `op-batcher` package and its subpackages. 

I could have missed something, but I think this PR actually moves closer to a unified definition of "pending". For example, `channel.pendingTransactions` appears to mean "transactions that are not yet confirmed on L1" (i.e., safe). Although it is a bit unclear to me whether a "pending" frame is _unsubmitted_ or _unsafe_--the latter is what we want. Edit: a pending frame is [unsubmitted](https://github.com/ethereum-optimism/optimism/blob/79355ae097bda6901248eb3cac6fa66620e07d01/op-batcher/batcher/channel_builder.go#L81). Let's discuss how to unify

joshklop replied:
Ah actually, `channelManager.pendingBlocks` needs to change its name. I'll probably keep reviewing and summarize the changes in another commit.

joshklop commented:
Closed in favor of #16828
